src/main/kotlin/com/nikitanazarov/coroutinestacks/CoroutineTraceForestBuilder.kt

```diff
@@ -138,34 +141,28 @@ private fun SuspendContextImpl.buildStackFrameGraph(
 ) {
     coroutineDataList.forEach { coroutineData ->
         var currentNode = rootValue
-
         val coroutineFrameItemLists: CoroutineFrameBuilder.Companion.CoroutineFrameItemLists
         try {
             coroutineFrameItemLists = CoroutineFrameBuilder.build(coroutineData, this) ?: return@forEach
         } catch (e : Exception) {
             return@forEach
         }
+
         coroutineFrameItemLists.frames.reversed().forEach { stackFrame ->
-            val xStackFrame = stackFrame.createFrame(debugProcess)
-            val isLibraryFrame = (xStackFrame as JVMStackFrameInfoProvider).isInLibraryContent
-            if (isLibraryFrame.not() or CoroutineStackPanelData.areLibraryFramesAllowed) {
+            val frameAllowed = if (areLibraryFramesAllowed) { true } else {
+                val xStackFrame = stackFrame.createFrame(debugProcess)
+                val isLibraryFrame = (xStackFrame as JVMStackFrameInfoProvider).isInLibraryContent
+                isLibraryFrame.not()
+            }
+            if (frameAllowed) {
                 val location = stackFrame.location
-                val child = currentNode.children[location]
-
-                if (child != null) {
-                    child.num++
-                    child.coroutinesActive += coroutineData.render()
-                    currentNode = child
-                } else {
-                    val node = Node(
-                        stackFrame,
-                        1,
-                        mutableMapOf(),
-                        coroutineData.render()
-                    )
-                    currentNode.children[location] = node
-                    currentNode = node
+                val child = currentNode.children.getOrPut(location) {
+                    Node(stackFrame, 0, mutableMapOf(), "")
                 }
+
+                child.num++
+                child.coroutinesActive += coroutineData.render()
```

Comment:
Now we have only one usage of the `render` function, so we can remove it and inline its body here